macos: change rules and improve detection of native SDK

kubkon opened the pull request:
Summary of breaking changes:
1. if we build on an Apple host (mainly macOS) and we target an Apple platform, we will try linking against the SDK if available
2. however, if we detect Nix, we ignore native SDK and instead rely fully on paths supplied by Nix via env vars

Now, the nitty-gritty:
In order to achieve 1. we have to apply this rule to all possible `Compilation.create` invocations which made it natural to move some of the existing code that is concerned with native paths detection into `Compilation.create`. And so, now detection of native Apple SDK is done within `Compilation.create`, as well as detection/unpacking of paths supplied by Nix. I think this bit of functionality really belongs in `LibcInstallation` struct, as it is a natural companion to autodetection logic of libc include paths that `LibcInstallation` provides. I haven't acted on this yet in this PR, so feel free to comment on the correctness of my assertion.

Detection of native libc include paths for Apple platforms is now done via `LibcInstallation` struct as it is done for other platforms - this got rid of a lot of special casing for Apple in `Compilation.zig` which is nice. The detection logic in `LibcInstallation` for Apple platforms relies on a two-tier logic: if SDK was detected, we use that, otherwise, we fallback to `clang` for detection. The latter is to mainly support the Nix environment. Anyhow, with this PR, `zig libc` now works on macOS (no more panicking). Instead we get this if the SDK is available:

```
# The directory that contains `stdlib.h`.
# On POSIX-like systems, include directories be found with: `cc -E -Wp,-v -xc /dev/null`
include_dir=/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.3.sdk/usr/include

# The system-specific include directory. May be the same as `include_dir`.
# On Windows it's the directory that includes `vcruntime.h`.
# On POSIX it's the directory that includes `sys/errno.h`.
sys_include_dir=/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.3.sdk/usr/include

# The directory that contains `crt1.o` or `crt2.o`.
# On POSIX, can be found with `cc -print-file-name=crt1.o`.
# Not needed when targeting MacOS.
crt_dir=

# The directory that contains `vcruntime.lib`.
# Only needed when targeting MSVC on Windows.
msvc_lib_dir=

# The directory that contains `kernel32.lib`.
# Only needed when targeting MSVC on Windows.
kernel32_lib_dir=

# The directory that contains `crtbeginS.o` and `crtendS.o`
# Only needed when targeting Haiku.
gcc_dir=

# The directory that contains system frameworks (Darwin only).
framework_dir=/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.3.sdk/System/Library/Frameworks
```

Closes #15024
Closes #14569

EDIT:
Closes #11631 

@haze marking #11631 as closed but I will ask you to verify it did indeed completely fix the issue for you once 0.11 lands, and is available via Nix.

EDIT2:
Closes https://github.com/ziglang/zig/issues/16118
Closes #15963

haze commented:
This should also close https://github.com/ziglang/zig/issues/11631 no?

andrewrk commented on src/Compilation.zig:
```diff
@@ -851,14 +851,60 @@ pub fn create(gpa: Allocator, options: InitOptions) !*Compilation {
+                std.log.warn("{s}", .{warning});
+            }
+
+            try clang_argv.ensureUnusedCapacity(paths.include_dirs.items.len * 2);
+            for (paths.include_dirs.items) |include_dir| {
+                clang_argv.appendAssumeCapacity("-isystem");
+                clang_argv.appendAssumeCapacity(include_dir);
+            }
+
+            try clang_argv.ensureUnusedCapacity(paths.framework_dirs.items.len * 2);
+            try framework_dirs.ensureUnusedCapacity(paths.framework_dirs.items.len);
+            for (paths.framework_dirs.items) |framework_dir| {
+                clang_argv.appendAssumeCapacity("-iframework");
+                clang_argv.appendAssumeCapacity(framework_dir);
+                framework_dirs.appendAssumeCapacity(framework_dir);
+            }
```
This logic should go in `addCCArgs`.

andrewrk commented on src/Compilation.zig:
```diff
@@ -851,14 +851,60 @@ pub fn create(gpa: Allocator, options: InitOptions) !*Compilation {
-                break :blk sdk.path;
-            } else {
-                break :blk null;
+        var clang_argv = std.ArrayList([]const u8).init(arena);
+        var lib_dirs = std.ArrayList([]const u8).init(arena);
+        var framework_dirs = std.ArrayList([]const u8).init(arena);
+        var rpath_list = std.ArrayList([]const u8).init(arena);
```
Can you try adding a commit that adds the needed state here (except clang_argv) to `LibCDirs` and moving the corresponding logic to `detectLibCIncludeDirs` or `getZigShippedLibCIncludeDirsDarwin`?

I think that might clean this up even more.

andrewrk reviewed (changes requested):
Looks great!

I see a couple more things that could be further improved, let me know what you think.

andrewrk commented on src/Compilation.zig:
```diff
@@ -1428,6 +1437,18 @@ pub fn create(gpa: Allocator, options: InitOptions) !*Compilation {
             system_libs.putAssumeCapacity(lib_name, options.system_lib_infos[i]);
         }
 
+        var lib_dirs = std.ArrayList([]const u8).init(arena);
+        try lib_dirs.appendSlice(options.lib_dirs);
+        try lib_dirs.appendSlice(libc_dirs.libc_lib_dir_list);
+
+        var framework_dirs = std.ArrayList([]const u8).init(arena);
+        try framework_dirs.appendSlice(options.framework_dirs);
+        try framework_dirs.appendSlice(libc_dirs.libc_framework_dir_list);
+
+        var rpath_list = std.ArrayList([]const u8).init(arena);
+        try rpath_list.appendSlice(options.rpath_list);
+        try rpath_list.appendSlice(libc_dirs.libc_rpath_list);
```
I don't think these should be added here. This information is available to whatever logic needs it via `link.File.options.libc_installation`

kubkon replied:
Hmm, all are needed by the linker, and framework paths are needed by the frontend (compiler) too. How else would you provide this info then?

andrewrk replied:
Both the linker and the frontend have access to `link.File.options.libc_installation`.

kubkon replied:
Oh, did I finally understand it that you mean we can recreate this info on request in whatever linker needs it?

andrewrk replied:
yes I believe that is correct :+1:

kubkon replied:
Took me long enough to get it...

andrewrk commented on src/Compilation.zig:
```diff
@@ -4898,27 +4917,80 @@ fn detectLibCFromLibCInstallation(arena: Allocator, target: Target, lci: *const
     }
 
+    if (want_native_paths) {
+        const paths = std.zig.system.NativePaths.detect(arena, target) catch |err| {
```
This function should not be called from `detectLibCFromLibCInstallation`. This function is intended to only extract libc information from the `LibCInstallation` and not inspect the host at all.

andrewrk reviewed (changes requested):
Getting closer. I still see some things where they don't belong. Apologies for the vague review feedback - hopefully knowing what *not* to do reduces the possibility space so it becomes more clear how to proceed.

kubkon commented:
> 

I'm not sure how to proceed from here for one reason: our libc detection mechanism so far assumed that libc is only a set of include dirs which is wrong. Depending on the host, in this case macos or any Apple platform, the linker needs also the knowledge of libc search dir paths, rpaths and framework dirs. All of those together with a set of include dirs make up an actually libc installation on a given host. With that in mind, I'm not sure how to go about addressing some of your feedback as it seems to be still based on the assumption that libc installation is only a set of include dirs. Apologies if it is not but it's difficult for me to work out exactly an action plan based on your latest feedback as it's a little too vague. Could you therefore clarify your feedback a bit more?

andrewrk commented:
Can you make sure that this solves #11066?

kubkon commented:
@andrewrk I haven't put the contents of `NativePaths` into `LibCDirs`, instead, I have made it an optional in `link.File.Options` for the linkers to have access to. Lemme know if you think this is heading in the right direction, and/or what else needs adjusting, etc.

andrewrk commented on lib/std/zig/system/NativePaths.zig:
```diff
@@ -188,6 +191,22 @@ fn deinitArray(array: *ArrayList([:0]u8)) void {
     array.deinit();
 }
 
+pub inline fn getIncludeDirs(self: *const NativePaths) []const [:0]const u8 {
+    return self.include_dirs.items;
+}
+
+pub inline fn getLibDirs(self: *const NativePaths) []const [:0]const u8 {
+    return self.lib_dirs.items;
+}
+
+pub inline fn getFrameworkDirs(self: *const NativePaths) []const [:0]const u8 {
+    return self.framework_dirs.items;
+}
+
+pub inline fn getRpaths(self: *const NativePaths) []const [:0]const u8 {
+    return self.rpaths.items;
+}
+
```
why `inline`?

why do these getters even exist?

kubkon replied:
It made sense to me instead of typing `paths.lib_dirs.items` everywhere but it's a matter of preference so I am happy to remove it.

andrewrk commented on src/Compilation.zig:
```diff
@@ -855,14 +855,31 @@ pub fn create(gpa: Allocator, options: InitOptions) !*Compilation {
-                break :blk sdk.path;
-            } else {
-                break :blk null;
+        const darwin_native = (comptime builtin.target.isDarwin()) and options.target.isDarwin() and
```
This is incorrect logic for `darwin_native`. When `options.is_native_os` is false, `darwin_native` must be false. Otherwise, it's going to look at the system sometimes when it should be cross-compiling. Zig should not rely on the Darwin SDK when cross-compiling.

kubkon replied:
Ah, fair point. The reason I re-introduced `comptime` check is because it then emits darwin-specific code that requires a host that spawn subprocesses which isn't always the case plus it should even exist unless it's a native OS AND Darwin. Can I make it more verbose and add `is_native_os` here but leave the `comptime` check? Or perhaps you have a nicer suggestion for this?

andrewrk commented on src/Compilation.zig:
```diff
@@ -855,14 +855,31 @@ pub fn create(gpa: Allocator, options: InitOptions) !*Compilation {
+            const paths = std.zig.system.NativePaths.detect(arena, options.target) catch |err| {
+                fatal("unable to detect native system paths: {s}", .{@errorName(err)});
+            };
+            for (paths.warnings.items) |warning| {
+                log.warn("{s}", .{warning});
```
can you give an example of what this looks like when it occurs?

kubkon replied:
I cannot, I copied the code that was already there. You may either ask the original author, or we can remove it.

kubkon commented on src/link.zig:
```diff
@@ -204,6 +205,8 @@ pub const Options = struct {
     version: ?std.SemanticVersion,
     compatibility_version: ?std.SemanticVersion,
     libc_installation: ?*const LibCInstallation,
+    // TODO figure out if we can make it part of LibCInstallation
+    native_paths: ?NativePaths = null,
```
Sure, but I will require assistance here as I don't yet know how to do that.

andrewrk commented:
I solved these problems by expanding our existing frontend abstractions that are already designed to handle this situation (and already do handle it on the other operating systems besides macOS). Namely, `std.zig.system.NativePaths` and `LibCInstallation`. I made those changes in #16058. Please test that branch and suggest any enhancements there.
